be/src/env/env_s3.cpp

```diff
@@ -66,33 +82,96 @@ StatusOr<std::unique_ptr<RandomAccessFile>> EnvS3::new_random_access_file(const
-    config.scheme = Aws::Http::Scheme::HTTP;
-    if (is_oss_path(namenode.c_str())) {
-        if (config::object_storage_access_key_id.empty() || config::object_storage_secret_access_key.empty() ||
-            config::object_storage_endpoint.empty()) {
-            return Status::RuntimeError(
-                    "Configuration object_storage_access_key_id, object_storage_secret_access_key "
-                    "and object_storage_endpoint is required for OSS.");
+
+    Status appendv(const Slice* data, size_t cnt) override {
+        for (size_t i = 0; i < cnt; i++) {
+            RETURN_IF_ERROR(append(data[i]));
         }
-        config.endpointOverride = get_endpoint_from_oss_bucket(config::object_storage_endpoint, &bucket);
+        return Status::OK();
+    }
+
+    Status pre_allocate(uint64_t size) override { return Status::NotSupported("OutputStreamAdapter::pre_allocate"); }
+
+    Status close() override { return _os->close(); }
+
+    // NOTE: unlike posix file, the file cannot be writen anymore after `flush`ed.
+    Status flush(FlushMode mode) override { return _os->close(); }
+
+    // NOTE: unlike posix file, the file cannot be writen anymore after `sync`ed.
+    Status sync() override { return _os->close(); }
+
+    uint64_t size() const { return _bytes_written; }
+
+    const std::string& filename() const override { return _name; }
+
+private:
+    std::unique_ptr<io::OutputStream> _os;
+    std::string _name;
+    uint64_t _bytes_written;
+};
+
+// TODO: client cache
+static std::shared_ptr<Aws::S3::S3Client> new_s3client(const S3URI& uri) {
+    Aws::Client::ClientConfiguration config;
+    config.scheme = Aws::Http::Scheme::HTTP; // TODO: use the schema in uri
+    if (!uri.endpoint().empty()) {
+        config.endpointOverride = uri.endpoint();
     } else if (!config::object_storage_endpoint.empty()) {
         config.endpointOverride = config::object_storage_endpoint;
     }
     config.maxConnections = config::object_storage_max_connection;
-    S3ObjectStore store(config);
-    RETURN_IF_ERROR(store.init(false));
-    ASSIGN_OR_RETURN(auto input_file, store.get_object(bucket, object));
-    return std::make_unique<RandomAccessFileWrapper>(std::move(input_file), path);
+    const auto& access_key_id = config::object_storage_access_key_id;
+    const auto& secret_access_key = config::object_storage_secret_access_key;
+    // TODO: support pass access key by URI?
+    if (!access_key_id.empty() && !secret_access_key.empty()) {
+        auto credentials = std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(access_key_id, secret_access_key);
+        return std::make_shared<Aws::S3::S3Client>(credentials, config);
+    } else {
+        // if not cred provided, we can use default cred in aws profile.
+        return std::make_shared<Aws::S3::S3Client>(config);
+    }
+}
+
+StatusOr<std::unique_ptr<RandomAccessFile>> EnvS3::new_random_access_file(const RandomAccessFileOptions& opts,
```

Comment:
i don't think it's necessary to create a new client for every file.

Reply:
I already added a TODO above.

Reply:
https://github.com/StarRocks/starrocks/issues/3794